rules/cloud/okta/okta_password_in_alternateid_field.yml

```diff
@@ -0,0 +1,43 @@
+title: Okta Password in AlternateID Field
+id: 91b76b84-8589-47aa-9605-c837583b82a9
+status: experimental
+description: |
+    Detects when a user has potentially entered their password into the
+    username field, which will cause the password to be retained in log files
+references:
+    - https://developer.okta.com/docs/reference/api/system-log/
+    - https://www.mitiga.io/blog/how-okta-passwords-can-be-compromised-uncovering-a-risk-to-user-data
+    - https://support.okta.com/help/s/article/What-special-characters-are-accepted-by-the-Okta-password
+author: kelnage
+date: 2023/04/03
+modified: 2023/04/03
+tags:
+    - attack.credential_access
+    - attack.t1552
+logsource:
+    product: okta
+    service: okta
+detection:
+    selection:
+        legacyeventtype: core.user_auth.login_failed
+        actor_alternateid|re|all:
+            # The following check for Okta's password complexity requirements
+            # You may need to adapt them based on your specific Okta config
+            - .*[a-z].*  # A lowercase character
+            - .*[A-Z].*  # An uppercase character
+            - .*\d.*     # A number
+            # A symbol/special character
+            - .*[\"!\#\$%\&'’\(\)\*\+,\-\./\\:;<=>\?@\[\]\^_`\{\|\}\~].*
```

Comment:
You cannot combine the `re` modifier like that. Instead, every regex expression must be expressed in a separate line and then combine them. Example:

```yml
- actor_alternateid|re: ...
- actor_alternateid|re: ...
- actor_alternateid|re: ...
```

Also, the regexes are super greedy in my opinion and the rule will change depending on the env password policy (if I understood your comment).

You could make it a placeholder rule instead. See [example](https://github.com/SigmaHQ/sigma/tree/master/rules-placeholder) and leave it to the user.

But I really dislike the regex in it's current form.

Reply:
I have implemented a similar rule for an environment. This can easily detected if you have e.g. a maximum length in user names and a minimum password length longer than that. Or if the naming schema of the user name is somewhat restricted. I think it's a great detection idea but as @nasbench already said, it has to be tuned to the specific environment.

Does okta always use an e-mail address as a login name? If so you could write a regex matching e-mail addresses and trigger if the regex does not match on the user name field. This does not have 100% coverage but would have 0% false positives.

Reply:
> You cannot combine the `re` modifier like that

In what context are you discussing here? Because I've tested this script against several pySigma backends and they all generate valid queries:

```
sigma convert -t splunk okta_password_in_alternateid_field.yml --without-pipeline
Parsing Sigma rules  [####################################]  100%
legacyeventtype="core.user_auth.login_failed"  NOT (actor_alternateid IN ("0oa*", "*@okta.com"))
| regex actor_alternateid=".*[a-z].*"
| regex actor_alternateid=".*[A-Z].*"
| regex actor_alternateid=".*\\d.*"
| regex actor_alternateid=".*[\\\"!\\#\\$%\\&'’\\(\\)\\*\\+,\\-\\./\\\\:;<=>\\?@\\[\\]\\^_`\\{\\|\\}\\~].*"
```

```
sigma convert -t loki okta_password_in_alternateid_field.yml 
Parsing Sigma rules  [####################################]  100%
{job=~".+"} | logfmt | legacyeventtype=`core.user_auth.login_failed` and actor_alternateid=~`.*[a-z].*` and actor_alternateid=~`.*[A-Z].*` and actor_alternateid=~`.*\d.*` and actor_alternateid=~".*[\\\"!\\#\\$%\\&'’\\(\\)\\*\\+,\\-\\./\\\\:;<=>\\?@\\[\\]\\^_`\\{\\|\\}\\~].*" and actor_alternateid!~`(?i)0oa.*` and actor_alternateid!~`(?i).*@okta\.com`
```

The alternate approach to such a valid password regex, used by the original blogpost, was using zero-width assertions: `(?=.*\d)(?=.*[a-z])(?=.*[A-Z])` - but that's using a regex feature that not all backends will necessarily support (I know Loki doesn't).

However, I do understand both your concerns about the requirement for tuning to a specific environment. @phantinuss yes, the _default_ format for [Okta usernames is an email address](https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-create-character-restriction.htm) - so I could go with something along the lines of filtering out `[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}` perhaps? Misses edge cases of some newer domains (i.e., those that are longer than 4 characters), but should be valid for most configurations.

I'd be happy to move this rule into the `rule-placeholder` folder if it fits there better. If we go down that route, would we have to have a single regex placeholder for the valid usernames?

Reply:
I would guess that most okta environments use email addresses. So going forward with the "username is not an email address" rule in my opinion would be fine. As a normal rule in the ruleset. We can add a false positives text that explains that if someone has a different user name scheme, they have to disable the rule (and optionally create their own).

I would increase the TLD part to something like `[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,10}`. It's also not complete but would take care of some newer TLDs. Regex should be case sensitive (at least it can be, if the backend supports it). So I would add the lower-case characters, just in case.

Also this regex is not RFC5322 compliant. But at least it is readable :D (and I don't see any obvious characters missing)

Reply:
Okay, I've had a go at updating the rule to take this alternative approach (good catch on the case-sensitivity bit @phantinuss - don't know what I was thinking! 😅). You are right that it isn't 100% accurate - but as you say, the fully valid regex is a bit of nightmare to include in the rule 😆

@nasbench does this approach to the regular expression for the rule look better to you?

Reply:
LGTM :)